src/community/oseo/oseo-core/src/main/java/org/geoserver/opensearch/eo/store/WorkspaceAwareDecoratingFeatureSource.java

```diff
@@ -0,0 +1,165 @@
+import org.geotools.data.simple.SimpleFeatureCollection;
+import org.geotools.data.simple.SimpleFeatureSource;
+import org.geotools.factory.CommonFactoryFinder;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.opengis.filter.Filter;
+import org.opengis.filter.FilterFactory2;
+import org.opengis.filter.identity.FeatureId;
+
+/** Adds a workspace filter to the queries/filters, if a workspace is set */
+public class WorkspaceAwareDecoratingFeatureSource extends DecoratingSimpleFeatureSource {
+    private WorkspaceInfo workspaceInfo;
+    private final JDBCOpenSearchAccess openSearchAccess;
+
+    static final FilterFactory2 FF = CommonFactoryFinder.getFilterFactory2();
+    public static String WORKSPACES_FIELD = "workspaces";
+
+    /**
+     * Constructor
+     *
+     * @param delegate the delegate feature source
+     * @param workspaceInfo the workspace info
+     * @param openSearchAccess the OpenSearchAccess
+     */
+    public WorkspaceAwareDecoratingFeatureSource(
+            SimpleFeatureSource delegate,
+            WorkspaceInfo workspaceInfo,
+            JDBCOpenSearchAccess openSearchAccess) {
+        super(delegate);
+        this.workspaceInfo = workspaceInfo;
+
+        this.openSearchAccess = openSearchAccess;
+    }
+
+    @Override
+    public ReferencedEnvelope getBounds(Query query) throws IOException {
+        query = appendWorkspaceToQuery(query);
+        return delegate.getBounds(query);
+    }
+
+    @Override
+    public int getCount(Query query) throws IOException {
+        query = appendWorkspaceToQuery(query);
+        return delegate.getCount(query);
+    }
+
+    @Override
+    public SimpleFeatureCollection getFeatures(Filter filter) throws IOException {
+        return delegate.getFeatures(appendWorkspaceToFilter(filter));
+    }
+
+    @Override
+    public SimpleFeatureCollection getFeatures(Query query) throws IOException {
+        query = appendWorkspaceToQuery(query);
+        return delegate.getFeatures(query);
+    }
+
+    private Set<FeatureId> getCollectionIdsForWorkspace() throws IOException {
+        String workspace = null;
+        Set<FeatureId> ids = new LinkedHashSet<>();
+        if (workspaceInfo != null) {
+            SimpleFeatureCollection idFeatureCollection = getIdFeatureCollection();
+            idFeatureCollection.accepts(f -> ids.add(f.getIdentifier()), null);
+        }
+        return ids;
+    }
+
+    private Set<String> getCollectionNamesForWorkspace() throws IOException {
+        String workspace = null;
+        Set<String> names = new LinkedHashSet<>();
+        if (workspaceInfo != null) {
+            SimpleFeatureCollection idFeatureCollection = getIdFeatureCollection();
+            idFeatureCollection.accepts(
+                    f -> names.add(f.getProperty("name").getValue().toString()), null);
+        }
+        return names;
+    }
+
+    private SimpleFeatureCollection getIdFeatureCollection() throws IOException {
```

Comment:
This one is not really returning something special for "id", one extracts both names and ids in fact. I'd go for ``getWorkspaceCollections``.